protocols/bizcoin/bizcoin.go

```diff
@@ -0,0 +1,819 @@
+
+		if bz.IsLeaf() {
+			return bz.startCommitmentPrepare()
+		}
+	case ROUND_COMMIT:
+		announcement = &BizCoinAnnounce{
+			TYPE:         ROUND_COMMIT,
+			Announcement: bz.commit.Announce(ann.Announcement),
+		}
+		dbg.Lvl3(bz.Name(), "BizCoin Handle Announcement COMMIT")
+
+		if bz.IsLeaf() {
+			return bz.startCommitmentCommit()
+		}
+	}
+
+	var err error
+	for _, tn := range bz.Children() {
+		err = bz.SendTo(tn, announcement)
+	}
+	return err
+}
+
+// startPrepareCommitment send the first commitment up the tree for the prepare
+// round.
+func (bz *BizCoin) startCommitmentPrepare() error {
+	cm := bz.prepare.CreateCommitment()
+	err := bz.SendTo(bz.Parent(), &BizCoinCommitment{TYPE: ROUND_PREPARE, Commitment: cm})
+	dbg.Lvl3(bz.Name(), "BizCoin Start Commitment PREPARE")
+	return err
+}
+
+// startCommitCommitment send the first commitment up the tree for the
+// commitment round.
+func (bz *BizCoin) startCommitmentCommit() error {
+	cm := bz.commit.CreateCommitment()
+
+	err := bz.SendTo(bz.Parent(), &BizCoinCommitment{TYPE: ROUND_COMMIT, Commitment: cm})
+	dbg.Lvl3(bz.Name(), "BizCoin Start Commitment COMMIT", err)
+	return err
+}
+
+// handle the arrival of a commitment
+func (bz *BizCoin) handleCommit(ann BizCoinCommitment) error {
+	var commitment *BizCoinCommitment
+	// store it and check if we have enough commitments
+	switch ann.TYPE {
+	case ROUND_PREPARE:
+		bz.tpcMut.Lock()
+		bz.tempPrepareCommit = append(bz.tempPrepareCommit, ann.Commitment)
+		if len(bz.tempPrepareCommit) < len(bz.Children()) {
+			bz.tpcMut.Unlock()
+			return nil
+		}
+		commit := bz.prepare.Commit(bz.tempPrepareCommit)
+		bz.tpcMut.Unlock()
+		if bz.IsRoot() {
+			dbg.Print(bz.Name(), "handle Commit PREPARE => WIll sTART Challenge !")
+			return bz.startChallengePrepare()
+		}
+		commitment = &BizCoinCommitment{
+			TYPE:       ROUND_PREPARE,
+			Commitment: commit,
+		}
+		dbg.Lvl3(bz.Name(), "BizCoin handle Commit PREPARE")
+	case ROUND_COMMIT:
+		bz.tccMut.Lock()
+		bz.tempCommitCommit = append(bz.tempCommitCommit, ann.Commitment)
+		if len(bz.tempCommitCommit) < len(bz.Children()) {
+			bz.tccMut.Unlock()
+			return nil
+		}
+		commit := bz.commit.Commit(bz.tempCommitCommit)
+		bz.tccMut.Unlock()
+		if bz.IsRoot() {
+			// do nothing
+			//	bz.startChallengeCommit()
+			// stop the processing of the round, wait the end of the "prepare"
+			// round. startChallengeCOmmit will be called then.
+			return nil
+		}
+		commitment = &BizCoinCommitment{
+			TYPE:       ROUND_COMMIT,
+			Commitment: commit,
+		}
+		dbg.Lvl3(bz.Name(), "BizCoin handle Commit COMMIT")
+	}
+	err := bz.SendTo(bz.Parent(), commitment)
+	dbg.Print(bz.Name(), "HandleCommit() sent to parent!")
+	return err
+}
+
+// startPrepareChallenge create the challenge and send its down the tree
+func (bz *BizCoin) startChallengePrepare() error {
+	// make the challenge out of it
+	trblock := bz.tempBlock
+	marshalled, err := json.Marshal(trblock)
```

Comment:
Yeah we just used the same functions that lefteris did using the old code. He is using that as a Hash function. That's not right I agree, but for the simulation that was sufficient.